### PR TITLE
fix: provide a sensible example for a privateca Root CA example

### LINKED ISSUE #630: privateca root ca example is invalid for a standard compliant root ca
### TL;DR

The sample in privateca/certificate_authority_basic/main.tf looks like it's a copy of the subordinate setup and not for the root.

### Expected behavior

Sample should be somewhat compliant to RFC 5280 and CA/B Baseline Requirements.

### Observed behavior

SAN on Root -> does not make any sense
pathLen on Root is not forbidden but according to the rfc not evaluated and not recommended by CA/B BR
extendedKeyUsage is forbidden by CA/B BR on a root 

### Terraform Configuration

```hcl
does not apply
```

### Terraform Version

```sh
does not apply
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by hoexter**
This one looks a lot like someone copied by accident the subordinate example out of `certificate_authority_subordinate/main.tf` as a root CA. Thus it contains a lot of values set which are outright invalid or not recommend for Root CA certficates if you consider RFC 5280 and CA/B Baseline Requirements as the standard to follow.

Also the subordinate example is a bit odd, e.g. configuring SAN on any kind of CA certificate doesn't make sense. And the resources examples there make use of the same pool name.

I tried to keep the lifetime setting, but set it to 99 years. That is probably a sensible value for a P(rivate)KI setup. For something public 10y or 15y are probably more sensible.

## Description

Fixes #630 

Note: If you are not associated with Google, open an issue for discussion before submitting a pull request.

## Checklist

**Readiness**

- [] Yes, **merge** this PR after it is approved
- [X] No, don't **merge** this PR after it is approved

**Testing**

- [ ] I have performed tests described in the [Contributing guide](https://github.com/terraform-google-modules/terraform-docs-samples/blob/main/CONTRIBUTING.md):

   - [ ] **[Tests](https://github.com/terraform-google-modules/terraform-docs-samples/blob/main/CONTRIBUTING.md#set-up-the-test-environment)** pass: `terraform apply`
   - [x] **[Lint](https://github.com/terraform-google-modules/terraform-docs-samples/blob/main/CONTRIBUTING.md#linting-and-formatting)** pass: `terraform fmt` check

--> this should get a test run somewhere, right now I don't have a test setup at hand to validate it against the API of the CAS

### CODE REVIEW

File: privateca/certificate_authority_basic/main.tf
Comment (iennae) [inline privateca/certificate_authority_basic/main.tf:45]: I think for the basic case this should be true; matching up to documentation here https://registry.terraform.io/providers/hashicorp/google/5.24.0/docs/resources/privateca_certificate_authority. 

For the subordinate it's false.

File: privateca/certificate_authority_basic/main.tf
Comment (iennae) [inline privateca/certificate_authority_basic/main.tf:53]: the original configuration here is just matching up with the example used in the main documentation I believe. I think your update here points out something important we should say explicitly even if we leave the configuration to match the existing example with a comment of something like

// lifetime for Root CA should be valid indefinitely so outside of 
// testing environments, update to a longer span of time

File: privateca/certificate_authority_subordinate/main.tf
Comment (iennae) [inline privateca/certificate_authority_subordinate/main.tf:60]: interesting; in the https://registry.terraform.io/providers/hashicorp/google/5.24.0/docs/resources/privateca_certificate_authority#example-usage---privateca-certificate-authority-subordinate example this uses the same pool name for both

File: privateca/certificate_authority_basic/main.tf
Comment (hoexter) [reply privateca/certificate_authority_basic/main.tf:45]: Plesae don't look at the provider configuration examples, those are equally wrong. Please take a look at either RFCs or the CA/B BR. E.g. here for the EKU values take a look at https://cabforum.org/uploads/CA-Browser-Forum-TLS-BRs-v2.0.2.pdf  7.1.2.1.2 Root CA Extensions -> It's "Extension extKeyUsage" "Presence MUST NOT".
For a subordinate the EKU values are a MUST if it's not empty aka unrestricted.

File: privateca/certificate_authority_subordinate/main.tf
Comment (hoexter) [reply privateca/certificate_authority_subordinate/main.tf:60]: I believe technically the Google Cloud CAS product allows you to have different types of certificates in a single pool. But usually that does not make a whole lot of sense. Because in practice you also configure policy restrictions on a pool for issuing (in addition to the runtime constrains encoded in the certificate which are evaluated during usage by the tls library) and those will be fundamentally different for root certificates and subordinates.
Maybe for sample code that is used with automated testing it's useful to just stuff all certs into a single pool, but that is irritating for people aiming at building a working implementation.

Otherwise I might got a wrong understanding of how the CAS product and its pools should be used for the past two years. ;)

File: privateca/certificate_authority_basic/main.tf
Comment (hoexter) [reply privateca/certificate_authority_basic/main.tf:53]: Fine for me either way. That's the least of my concern with the current documentation, because it's just a setting that is not helpful to run a CA in practice, but otherwise a correct.

File: privateca/certificate_authority_basic/main.tf
Comment (hoexter) [reply privateca/certificate_authority_basic/main.tf:45]: Just to mention the obvious I'm not a PKIX expert myself, but inside Google you also have wizards like Ryan Sleevi, though trivial stuff like this might be not a good target for a wizard. The people who developed the CAS product should be also fine to consult with for the details.

File: privateca/certificate_authority_subordinate/main.tf
Comment (hoexter) [reply privateca/certificate_authority_subordinate/main.tf:60]: Due to coincidence I just saw a new test case in the newly released TF provider which does exactly that: Setup a pool for the root and setup a pool for the subordinate:
https://github.com/hashicorp/terraform-provider-google-beta/pull/7197/files#diff-19706c53088b05bc2bc72f7bd03ee2eacbe6d12052408d3e905f150d2f3baafdR320

But it also configures EKU values on the root certificate, which is wrong there as well. I assume that is just wrong sample code that gets copied and pasted around now everywhere now.

**Comment by Sita04** [comment]
@msampathkumar Can you help add @pmansour to the list of reviewers?

File: privateca/certificate_authority_basic/main.tf
Comment (pmansour) [reply privateca/certificate_authority_basic/main.tf:45]: Agreed with @hoexter: by default, we probably don't want any EKUs here.

EKUs can appear on a root CA, but they have a different meaning there -- specifically, they function to limit the allowed set of EKUs on all intermediate CAs and leaf certificates in that chain. That's probably not something most people want to do at their root-level CA.

File: privateca/certificate_authority_basic/main.tf
Comment (pmansour) [reply privateca/certificate_authority_basic/main.tf:53]: I don't think we want to leave the impression that root CAs last forever (or 99 years). We want most people to explicitly recognize that root CAs will need to be rotated (whether due to expiry, compromise, or just good hygiene) and plan for that.

Having said that, 86,400 seconds is only 1 day which really doesn't make sense for a root CA.

A more reasonable default lifetime for a root CA is 10 years (or even lower, e.g. 5 years). This is the same default we use for a root CA in gcloud: https://cloud.google.com/sdk/gcloud/reference/privateca/roots/create#--validity

So I'd suggest we use
```
lifetime = "${10 * 365 * 24 * 3600}s"
```

File: privateca/certificate_authority_basic/main.tf
Comment (pmansour) [reply privateca/certificate_authority_basic/main.tf:45]: Btw, in this case I'd suggest dropping the `extended_key_usage { .. }` block completely, rather than including it with only false values.

File: privateca/certificate_authority_subordinate/main.tf
Comment (pmansour) [inline privateca/certificate_authority_subordinate/main.tf:33]: could you also remove the `extended_key_usage { .. }` block here as well?

File: privateca/certificate_authority_subordinate/main.tf
Comment (pmansour) [inline privateca/certificate_authority_subordinate/main.tf:53]: same as above, please update the comment and value to be 10 years, and avoid the implication of indefinite validity

File: privateca/certificate_authority_subordinate/main.tf
Comment (pmansour) [reply privateca/certificate_authority_subordinate/main.tf:60]: Agreed with @hoexter: in almost all cases, roots and subordinates should be in separate pools. The rationale behind CA Pools is that all CAs in the same pool should be completely interchangeable. So having 1 pool for the root CA(s) and another for the subordinate CA(s) is the way to go.

File: privateca/certificate_authority_subordinate/main.tf
Comment (pmansour) [inline privateca/certificate_authority_subordinate/main.tf:57]: let's also replace `"default"` here with `"sub_ca"`, similar to what you did for the root

File: privateca/certificate_authority_subordinate/main.tf
Comment (pmansour) [inline privateca/certificate_authority_subordinate/main.tf:63]: _(sorry for the misplaced comment -- I can't figure out how to comment on the actual lines I want to mention, which weren't modified here)_

Could you also make the following changes for the sub CA:
- Remove the `subject_alt_name`
- Get rid of all `base_key_usage` values except for `cert_sign` and `crl_sign`.
- Get rid of the `extended_key_usage` block entirely
- Update the lifetime from 1 day (86400s) to 5 years (`"{5 * 365 * 24 * 3600}"`)
- Update the signing key algorithm to use `RSA_PKCS1_2048_SHA256` (more efficient for use in a sub CA)

**Review by pmansour — CHANGES_REQUESTED** [review]
Great callouts, thanks for helping to fix this! The values in this sample were probably copied over from somewhere without much thought about reasonable defaults.

- Peter from the CAS eng team

File: privateca/certificate_authority_basic/main.tf
Comment (hoexter) [reply privateca/certificate_authority_basic/main.tf:45]: Currently the language server tells me that the provider expects the `extended_key_usage {}` block to be present, maybe it can be empty.

File: privateca/certificate_authority_subordinate/main.tf
Comment (hoexter) [reply privateca/certificate_authority_subordinate/main.tf:62]: done, also took the freedom to replace HashiCorp with ACME
eku block is empty, someone has to check what the provider actually makes out of it

File: privateca/certificate_authority_basic/main.tf
Comment (hoexter) [reply privateca/certificate_authority_basic/main.tf]: done, albeit my personal opinion is slightly different on that topic because if I ran a private key infrastructure I usually move a CA out of service by removing it from the ca store of my devices and by no longer issuing certificates for it. Though someone might also use this product for more distributed use cases with less control. :)
In general trying to set a good example with short lifetimes to train for crypto agility is likely a good thing.

File: privateca/certificate_authority_subordinate/main.tf
Comment (hoexter) [reply privateca/certificate_authority_subordinate/main.tf]: done

File: privateca/certificate_authority_subordinate/main.tf
Comment (hoexter) [reply privateca/certificate_authority_subordinate/main.tf:62]: https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/products/privateca/CertificateAuthority.yaml#L404 there it's requiring the EKU block

**Comment by msampathkumar** [comment]
This PR has been inactive for two months. If this inactivity continues for another two weeks, I will close the request.

**Comment by hoexter** [comment]
@msampathkumar I updated the PR long ago with the proposed changes by @pmansour. Thus it's just waiting for a review by you/your team. :)

File: privateca/certificate_authority_subordinate/main.tf
Comment (glasnt) [reply privateca/certificate_authority_subordinate/main.tf:59]: This change is causing the tests to fail. This is due to the comment above this value, where this pool needs to already exist. 

I've created https://github.com/terraform-google-modules/terraform-docs-samples/pull/748 to support this change, which needs to be approved and merged before this PR will pass testing.

Comment (glasnt) [reply privateca/certificate_authority_subordinate/main.tf:59]: Tests now pass with this change.

**Comment by hoexter** [comment]
Can I do anything else to help you to review this PR and get it a state you can merge?

**Comment by glasnt** [comment]
I haven't been able to find a subject matter expert on this, but looking at the changes made, it seems reasonable. I'm not an expert on RFC5280, but removing some of the extra fields appears to match what modules do (e.g.. csa-certificate-authority-service)

**Review by glasnt — APPROVED** [review]

**Pull request merged**
